JS: model fastify

esbena opened the pull request:
Yet another minimal security-related model of a HTTP server implementation. The implementation is mostly boiler-plate for this kind of model. It would be nice if the already extensive HTTP.qll could recude the boiler-plate even more, but that can be done in a separate PR.

I am evaluating this on a collection of fastify dependents.

Addresses <https://github.com/github/codeql/issues/3435>.

asgerf commented on javascript/ql/src/semmle/javascript/frameworks/Fastify.qll:
```diff
@@ -0,0 +1,249 @@
+  /**
+   * A function used as a Fastify route handler.
+   *
+   * By default, only handlers installed by an Fastify route setup are recognized,
```
```suggestion
   * By default, only handlers installed by a Fastify route setup are recognized,
```

asgerf commented on javascript/ql/src/semmle/javascript/frameworks/Fastify.qll:
```diff
@@ -0,0 +1,249 @@
+    /**
+     * Gets the parameter of the route handler that contains the request object.
+     */
+    abstract SimpleParameter getRequestParameter();
```
Why restrict this to `SimpleParameter`?

esbena commented:
[Evaluation](https://git.semmle.com/esben/dist-compare-reports/tree/js/fastify_1589969124846): nothing of interest on a set of relatively small applications.

asgerf commented on javascript/ql/src/semmle/javascript/frameworks/Fastify.qll:
```diff
@@ -0,0 +1,249 @@
+    string kind;
+
+    RequestInputAccess() {
+      exists(string name | this.(DataFlow::PropRead).accesses(rh.getARequestExpr().flow(), name) |
```
```suggestion
      exists(string name | this = rh.getARequestSource().ref().getAPropertyRead(name) |
```

asgerf commented on javascript/ql/src/semmle/javascript/frameworks/Fastify.qll:
```diff
@@ -0,0 +1,249 @@
+
+    override RouteHandler getRouteHandler() { result = rh }
+
+    override string getKind() { result = kind }
```
Could you override `RemoteFlowSource.isUserControlledObject` as well? Otherwise we won't get much coverage for NoSQL injection.

asgerf commented on javascript/ql/src/semmle/javascript/frameworks/Fastify.qll:
```diff
@@ -0,0 +1,249 @@
+    RouteHandler rh;
+
+    RequestHeaderAccess() {
+      exists(DataFlow::PropRead headers |
+        headers.accesses(rh.getARequestExpr().flow(), "headers") and
+        this = headers.getAPropertyRead()
+      )
```
```suggestion
      this = rh.getARequestSource().ref().getAPropertyRead("headers").getAPropertyRead()
```

asgerf commented on javascript/ql/src/semmle/javascript/frameworks/Fastify.qll:
```diff
@@ -0,0 +1,249 @@
+    RouteHandler rh;
+
+    ResponseSendArgument() {
+      exists(MethodCallExpr mce |
+        mce.calls(rh.getAResponseExpr(), "send") and
+        this = mce.getArgument(0)
+      )
```
```suggestion
      this = rh.getAResponseSource().ref().getAMethodCall("send").getArgument(0).asExpr()
```

asgerf commented on javascript/ql/src/semmle/javascript/frameworks/Fastify.qll:
```diff
@@ -0,0 +1,249 @@
+      exists(Function f |
+        f = rh.(DataFlow::FunctionNode).getFunction() and
+        f.isAsync() and
+        f.getAReturnedExpr() = this
```
Hm, presumably you could have a non-async function that just returns a promise? (unless they use nasty reflection).

For the benefit of the open redirect query, it would be best to just omit the `isAsync` requirement.

asgerf commented on javascript/ql/src/semmle/javascript/frameworks/Fastify.qll:
```diff
@@ -0,0 +1,249 @@
+  private class RedirectInvocation extends HTTP::RedirectInvocation, MethodCallExpr {
+    RouteHandler rh;
+
+    RedirectInvocation() { this.calls(rh.getAResponseExpr(), "redirect") }
```
```suggestion
    RedirectInvocation() { this = rh.getAResponseSource().ref().getAMethodCall("redirect").asExpr() }
```

asgerf commented on javascript/ql/src/semmle/javascript/frameworks/Fastify.qll:
```diff
@@ -0,0 +1,249 @@
+  /**
+   * An invocation that sets a single header of the HTTP response.
+   */
+  private class SetOneHeader extends HTTP::Servers::StandardHeaderDefinition {
+    RouteHandler rh;
+
+    SetOneHeader() {
+      astNode.calls(rh.getAResponseExpr(), "header") and
+      astNode.getNumArgument() = 2
+    }
```
```suggestion
  private class SetOneHeader extends HTTP::Servers::StandardHeaderDefinition, DataFlow::MethodCallNode {
    RouteHandler rh;

    SetOneHeader() {
      this = rh.getAResponseSource().ref().getAMethodCall("header") and
      getNumArgument() = 2
    }
```

asgerf commented on javascript/ql/src/semmle/javascript/frameworks/Fastify.qll:
```diff
@@ -0,0 +1,249 @@
+    RouteHandler rh;
+
+    SetMultipleHeaders() {
+      this.calls(rh.getAResponseExpr().flow(), "headers") and
+      this.getNumArgument() = 1
```
```suggestion
      this = rh.getAResponseSource().ref().getAMethodCall("headers") and
      this.getNumArgument() = 1
```

esbena commented:
> I'd like it if we could move towards the SourceNode chainable API over the calls/accesses predicates, and certainly away from Expr-based API when there is a data flow alternative (even though it's slightly less urgent after the recent PR).

I agree. I should have mentioned that this was a deliberate choice to match the existing code, I am sorry that I made you make the many suggestions. I have addressed all of the comments, except the one about user-controlled objects. I'll dig into how that works for this library...

esbena commented:
A simple `isUserControlledObject` has been implemented. Additional work items have been recorded in https://github.com/github/codeql-javascript-team/issues/128.

asgerf commented on javascript/ql/src/semmle/javascript/frameworks/Fastify.qll:
```diff
@@ -132,7 +132,9 @@ module Fastify {
     RequestInputAccess() {
-      exists(string name | this.(DataFlow::PropRead).accesses(rh.getARequestExpr().flow(), name) |
+      exists(DataFlow::PropRead read, string name |
+        this = read and read = rh.getARequestSource().ref().getAPropertyRead(name)
```
Is there a need for the `read` variable here?

esbena replied:
good spot.

esbena commented:
The dataflow-based [evaluation](https://git.semmle.com/esben/dist-compare-reports/tree/js/fastify_1589978959859) is equally boring.

esbena commented:
All comments have been addressed.

Ethan-Arrowood commented:
This is very nice work. The Fastify team appreciates it @esbena . If you ever need our input please feel free to reach out.

mcollina commented:
Thanks @esbena!

esbena commented:
Rebased with updated expected output.
